[PyTorch] Make RecordFunctionCallback::should_run_ a function pointer

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48629 [PyTorch] Use plain old function pointer for RecordFunctionCallback
* #48620 [PyTorch] remove convenience RecordFunctionCallback interface
* #48286 [PyTorch] Inline RecordFunctionCallback::shouldRun
* #47550 [PyTorch] Lazily construct guts of RecordFunction
* **#48274 [PyTorch] Make RecordFunctionCallback::should_run_ a function pointer**

The std::function-ness of it was used only for tests. (std::function is huge at 32 bytes, and not particularly efficient.)

Differential Revision: [D25102077](https://our.internmc.facebook.com/intern/diff/D25102077/)

swolchok committed:
[PyTorch] Make RecordFunctionCallback::should_run_ a function pointer

The std::function-ness of it was used only for tests. (std::function is huge at 32 bytes, and not particularly efficient.)

Differential Revision: [D25102077](https://our.internmc.facebook.com/intern/diff/D25102077/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make RecordFunctionCallback::should_run_ a function pointer"


The std::function-ness of it was used only for tests. (std::function is huge at 32 bytes, and not particularly efficient.)

Differential Revision: [D25102077](https://our.internmc.facebook.com/intern/diff/D25102077/)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Please make gods of clang-format happy

swolchok committed:
take a guess at clang-format fix on "[PyTorch] Make RecordFunctionCallback::should_run_ a function pointer"


The std::function-ness of it was used only for tests. (std::function is huge at 32 bytes, and not particularly efficient.)

Differential Revision: [D25102077](https://our.internmc.facebook.com/intern/diff/D25102077/)

[ghstack-poisoned]

swolchok committed:
clang-format on "[PyTorch] Make RecordFunctionCallback::should_run_ a function pointer"


The std::function-ness of it was used only for tests. (std::function is huge at 32 bytes, and not particularly efficient.)

Differential Revision: [D25102077](https://our.internmc.facebook.com/intern/diff/D25102077/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make RecordFunctionCallback::should_run_ a function pointer"


The std::function-ness of it was used only for tests. (std::function is huge at 32 bytes, and not particularly efficient.)

Differential Revision: [D25102077](https://our.internmc.facebook.com/intern/diff/D25102077/)

[ghstack-poisoned]
